plugins/newsdownloader.koplugin/feed_source.lua

```diff
@@ -0,0 +1,403 @@
+    else
+        if not ok then
+            error("(Reason: Failed to download feed document)")
+        else
+            error("(Reason: Error during feed document deserialization)")
+        end
+    end
+end
+
+-- Supply this method with the XML document returned by the feed,
+-- and it will initialized the document by extracting the feed title,
+-- feed items, and items count.
+function FeedSource:initializeDocument(document)
+    local feed_title
+    local feed_items
+    local total_items
+
+    local ok = pcall(function()
+            return self:getFeedType(
+                document,
+                function()
+                    -- RSS callback
+                    feed_title = util.htmlEntitiesToUtf8(document.rss.channel.title)
+                    feed_items = document.rss.channel.item
+                    total_items = #document.rss.channel.item
+                end,
+                function()
+                    -- Atom callback
+                    feed_title = FeedSource:getFeedTitle(document.feed.title)
+                    feed_items = document.feed.entry
+                    total_items = #document.feed.entry
+                end
+            )
+    end)
+
+    if ok then
+        document.title = feed_title
+        document.items = feed_items
+        document.total_items = total_items
+        return document
+    else
+        error("Could not initialize feed document")
+    end
+end
+
+function FeedSource:getItemsContent(feed, progress_callback, error_callback)
+    local limit = tonumber(feed.config.limit)
+    local total_items = (limit == 0) and
+        feed.document.total_items or
+        limit
+    local initialized_feed_items = {}
+    -- Download each ite0m in the feed
+    for index, item in pairs(feed.document.items) do
+        -- If limit has been met, stop downloading feed.
+        if limit ~= 0 and index - 1 == limit then
+            break
+        end
+        -- Display feedback to user.
+        progress_callback(T(
+            _("Getting item %1 of %2 from %3"),
+            index,
+            total_items,
+            feed.document.title
+        ))
+        -- Download the article's HTML.
+        local ok, response = pcall(function()
+               return self:initializeItemHtml(
+                    feed,
+                    self:getItemHtml(
+                        item,
+                        feed.config.download_full_article
+                    )
+                )
+        end)
+
+        -- Add the result to our table, or send a
+        -- result to the error callback.
+        if ok then
+            table.insert(
+                initialized_feed_items,
+                {
+                    html = response.html,
+                    images = response.images,
+                    item_slug = FeedSource:getItemTitleWithDate(item),
+                    item_title = item.title,
+                    md5 = md5(item.title),
+                    feed_title = feed.document.title
```

Comment:
We like trailing commas. That way you don't have to think about stupid things like commas when adding a new line, and on a diff it doesn't show up with misleading changed lines (in some instances).